src/main/java/net/neoforged/neoforge/client/event/EntityRenderersEvent.java

```diff
@@ -195,43 +196,59 @@ public EntityRendererProvider.Context getContext() {
 
         @ApiStatus.Internal
-        public CreateSkullModels(ImmutableMap.Builder<Type, SkullModelBase> builder, EntityModelSet entityModelSet) {
+        public CreateSkullModels(ImmutableMap.Builder<SkullBlock.Type, Function<EntityModelSet, SkullModelBase>> builder) {
             this.builder = builder;
-            this.entityModelSet = entityModelSet;
         }
 
         /**
-         * {@return the set of entity models}
+         * Registers a {@link SkullModel} for a skull block with the given {@link SkullBlock.Type}.
+         *
+         * @param type          a unique skull type; an exception will be thrown later if multiple mods (including vanilla)
+         *                      register models for the same type
+         * @param layerLocation the key that identifies the {@link LayerDefinition} used by the model
+         */
+        public void registerSkullModel(SkullBlock.Type type, ModelLayerLocation layerLocation) {
+            this.registerSkullModel(type, layerLocation, SkullModel::new);
+        }
+
+        /**
+         * Registers the entity model for a skull block with the given {@link SkullBlock.Type}.
+         *
+         * @param type          a unique skull type; an exception will be thrown later if multiple mods (including vanilla)
+         *                      register models for the same type
+         * @param layerLocation the key that identifies the {@link LayerDefinition} used by the model
+         * @param factory       the factory to create the skull model instance, taking in the root {@link ModelPart} and
+         *                      returning the model.
          */
-        public EntityModelSet getEntityModelSet() {
-            return entityModelSet;
+        public void registerSkullModel(SkullBlock.Type type, ModelLayerLocation layerLocation, Function<ModelPart, SkullModelBase> factory) {
+            this.registerSkullModel(type, modelSet -> factory.apply(modelSet.bakeLayer(layerLocation)));
         }
 
         /**
-         * Registers the constructor for a skull block with the given {@link SkullBlock.Type}.
-         * These will be inserted into the maps used by the item, entity, and block model renderers at the appropriate
-         * time.
+         * Registers the entity model for a skull block with the given {@link SkullBlock.Type}.
          *
-         * @param type  a unique skull type; an exception will be thrown later if multiple mods (including vanilla)
-         *              register models for the same type
-         * @param model the skull model instance. A typical implementation will simply bake a model using
-         *              {@link EntityModelSet#bakeLayer(ModelLayerLocation)} and pass it to the constructor for
-         *              {@link SkullModel}.
+         * @param type    a unique skull type; an exception will be thrown later if multiple mods (including vanilla)
```

Comment:
I would prefer if this were changed to fail fast (i.e. `putIfAbsent() != null -> throw`) by using a `HashMap` like the layer definitions do instead of relying on the immutable map builder throwing later.